admin/angel/Core.c

```diff
@@ -105,9 +110,33 @@ struct Context
     struct EventBase* base;
     struct NetCore* nc;
     struct IpTunnel* ipTunnel;
+#ifndef win32
+    struct FileNo_admin* fileno;
+#endif
     Identity
 };
 
+#ifndef win32
+static void onFileNoReceived(void* vcontext, enum FileNo_Type type, int fileno)
+{
+    struct Context* ctx = Identity_check((struct Context*) vcontext);
+    struct Jmp jmp;
+    Jmp_try(jmp) {
+        struct Pipe* p = Pipe_forFiles(fileno, fileno, ctx->base, &jmp.handler, ctx->alloc);
+        p->logger = ctx->logger;
+        if (type == FileNo_Type_ANDROID) {
+            struct AndroidWrapper* aw = AndroidWrapper_new(ctx->alloc, ctx->logger);
+            Iface_plumb(&aw->externalIf, &p->iface);
+            Iface_plumb(&aw->internalIf, &ctx->nc->tunAdapt->tunIf);
+        } else {
+            Iface_plumb(&p->iface, &ctx->nc->tunAdapt->tunIf);
+        }
+    } Jmp_catch {
+        Log_warn(ctx->logger, "Failed to configure fileno [%s]", jmp.message);
+    }
+}
+#endif
```

Comment:
This is wrong, FileNo_admin_new() should not take this callback, instead it should simply return the numeric representation of the file number to the caller when FileNo_import() is called, that way it is more generic than just important a fileno for the tun.
Once the fileno is imported into the cjdns process, *then* it should be given by Configurator to Core_initTunnel() which will do the logic which is seen here.